Error while checking build/O2/fullCI_slc9 for a46701def8cfcdb7f6e38ee3d219878a4d4b8fa8 at 2025-07-09 09:24:
```
## sw/BUILD/4ca51307b106213ce7ad18dea3427170d8d6aafc-latest/log
/sw/slc9_x86-64/YODA/yoda-2.1.0-6/include/YODA/highfive/H5Object.hpp:13:10: fatal error: H5Ipublic.h: No such file or directory
/sw/slc9_x86-64/YODA/yoda-2.1.0-6/include/YODA/highfive/H5Object.hpp:13:10: fatal error: H5Ipublic.h: No such file or directory
make[3]: *** [Makefile:619: libRivetCore_la-AnalysisLoader.lo] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: *** [Makefile:563: libRivetCore_la-Run.lo] Error 1
/sw/slc9_x86-64/YODA/yoda-2.1.0-6/include/YODA/highfive/H5Object.hpp:13:10: fatal error: H5Ipublic.h: No such file or directory
/sw/slc9_x86-64/YODA/yoda-2.1.0-6/include/YODA/highfive/H5Object.hpp:13:10: fatal error: H5Ipublic.h: No such file or directory
make[3]: *** [Makefile:612: libRivetCore_la-Analysis.lo] Error 1
make[3]: *** [Makefile:633: libRivetCore_la-AnalysisHandler.lo] Error 1
make[2]: *** [Makefile:659: all-recursive] Error 1
make[1]: *** [Makefile:567: all-recursive] Error 1
make: *** [Makefile:578: all-recursive] Error 1
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14493/a46701def8cfcdb7f6e38ee3d219878a4d4b8fa8/build_O2_fullCI_slc9/pretty.html).

--- a/Framework/Core/src/TopologyPolicy.cxx
+++ b/Framework/Core/src/TopologyPolicy.cxx
@@ -147,7 +147,7 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
   if (!isAExpendable && !isBExpendable) {
     bool sporadic = sporadicDataDeps(a, b);
     if (sporadic) {
-      O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "false. Neither %s nor %s are expendable. However the former has sporadic inputs so we sort it after.",
+      O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "true. Neither %s nor %s are expendable. However the former has sporadic inputs so we sort it after.",
                       a.name.c_str(), b.name.c_str());
       return true;
     }
@@ -159,7 +159,7 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
   if (isAExpendable && isBExpendable) {
     bool sporadic = sporadicDataDeps(a, b);
     if (sporadic) {
-      O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "false. Both %s and %s are expendable. However the former has sporadic inputs, so we sort it after.",
+      O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "true. Both %s and %s are expendable. However the former has sporadic inputs, so we sort it after.",
                       a.name.c_str(), b.name.c_str());
       return true;
     }
@@ -172,7 +172,7 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
   if (isAExpendable && bResilient) {
     bool sporadic = sporadicDataDeps(a, b);
     if (sporadic) {
-      O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "false. %s is expendable but %s is resilient, however the former also has sporadic inputs, so we sort it after.",
+      O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "true. %s is expendable but %s is resilient, however the former also has sporadic inputs, so we sort it after.",
                       a.name.c_str(), b.name.c_str());
       return true;
     }
@@ -188,8 +188,6 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
                     a.name.c_str(), hasDependency ? "There is however an inverse dependency" : "No inverse dependency", b.name.c_str(), a.name.c_str(),
                     !hasDependency ? "true" : "false");
     if (!hasDependency) {
-      O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "%s is expendable. There is however an inverse dependecy from %s to %s => true.",
-                      a.name.c_str(), b.name.c_str(), a.name.c_str());
       return true;
     }
     bool sporadic = sporadicDataDeps(a, b);
@@ -201,13 +199,6 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
     O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "%s is expendable. No inverse dependency from %s to %s => false.",
                     a.name.c_str(), b.name.c_str(), a.name.c_str());
     return false;
-  }
-  // b is expendable and a is not. We are fine with no dependency.
-  bool sporadic = sporadicDataDeps(a, b);
-  if (sporadic) {
-    O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "false. %s is expendable but %s is not. However the former has an sporadic input => true.",
-                    b.name.c_str(), a.name.c_str());
-    return true;
   }
   // b is expendable and a is not. We are fine with no dependency.
   O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "false. %s is expendable but %s is not. No need to add an unneeded dependency.",
@@ -270,6 +261,12 @@ TopologyPolicy::DependencyChecker TopologyPolicyHelpers::alwaysDependent()
                       hasDependency ? "true" : "false", dependent.name.c_str(), hasDependency ? "has" : "has not", ancestor.name.c_str());
       return hasDependency;
     }
+
+    if (sporadicDataDeps(ancestor, dependent)) {
+      O2_SIGNPOST_END(topology, sid, "alwaysDependent", "false. Dependent %s is an output proxy and ancestor %s has sporadic inputs", dependent.name.c_str(), ancestor.name.c_str());
+      return false;
+    }
+
     O2_SIGNPOST_END(topology, sid, "alwaysDependent", "true by default. Ancestor %s is not an output proxy.", ancestor.name.c_str());
     return true;
   };

--- a/Framework/Core/src/TopologyPolicy.cxx
+++ b/Framework/Core/src/TopologyPolicy.cxx
@@ -142,11 +142,14 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
   bool isBExpendable = std::find_if(b.labels.begin(), b.labels.end(), checkExpendable) != b.labels.end();
   bool isAExpendable = std::find_if(a.labels.begin(), a.labels.end(), checkExpendable) != a.labels.end();
   bool bResilient = std::find_if(b.labels.begin(), b.labels.end(), checkResilient) != b.labels.end();
+  const std::regex matcher(".*output-proxy.*");
+  std::cmatch m;
+  bool isBOutputProxy = std::regex_match(b.name.data(), m, matcher);
 
   // If none is expendable. We simply return false and sort as usual.
   if (!isAExpendable && !isBExpendable) {
     bool sporadic = sporadicDataDeps(a, b);
-    if (sporadic) {
+    if (sporadic && !isBOutputProxy) {
       O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "true. Neither %s nor %s are expendable. However the former has sporadic inputs so we sort it after.",
                       a.name.c_str(), b.name.c_str());
       return true;
@@ -158,7 +161,7 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
   // If both are expendable. We return false and sort as usual.
   if (isAExpendable && isBExpendable) {
     bool sporadic = sporadicDataDeps(a, b);
-    if (sporadic) {
+    if (sporadic && !isBOutputProxy) {
       O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "true. Both %s and %s are expendable. However the former has sporadic inputs, so we sort it after.",
                       a.name.c_str(), b.name.c_str());
       return true;
@@ -171,7 +174,7 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
   // If a is expendable but b is resilient, we can keep the same order.
   if (isAExpendable && bResilient) {
     bool sporadic = sporadicDataDeps(a, b);
-    if (sporadic) {
+    if (sporadic && !isBOutputProxy) {
       O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "true. %s is expendable but %s is resilient, however the former also has sporadic inputs, so we sort it after.",
                       a.name.c_str(), b.name.c_str());
       return true;
@@ -191,7 +194,7 @@ bool expendableDataDeps(DataProcessorSpec const& a, DataProcessorSpec const& b)
       return true;
     }
     bool sporadic = sporadicDataDeps(a, b);
-    if (sporadic) {
+    if (sporadic && !isBOutputProxy) {
       O2_SIGNPOST_END(topology, sid, "expendableDataDeps", "%s is expendable. No inverse dependency from %s to %s. However the former has an occasioanl input => true.",
                       a.name.c_str(), b.name.c_str(), a.name.c_str());
       return true;
@@ -260,11 +263,6 @@ TopologyPolicy::DependencyChecker TopologyPolicyHelpers::alwaysDependent()
       O2_SIGNPOST_END(topology, sid, "alwaysDependent", "%s. Dependent %s %s a dependency on ancestor %s.",
                       hasDependency ? "true" : "false", dependent.name.c_str(), hasDependency ? "has" : "has not", ancestor.name.c_str());
       return hasDependency;
-    }
-
-    if (sporadicDataDeps(ancestor, dependent)) {
-      O2_SIGNPOST_END(topology, sid, "alwaysDependent", "false. Dependent %s is an output proxy and ancestor %s has sporadic inputs", dependent.name.c_str(), ancestor.name.c_str());
-      return false;
     }
 
     O2_SIGNPOST_END(topology, sid, "alwaysDependent", "true by default. Ancestor %s is not an output proxy.", ancestor.name.c_str());

